lib/multi.c

```diff
@@ -2373,6 +2373,14 @@ static CURLMcode multi_runsingle(struct Curl_multi *multi,
    * again during processing, triggering a re-run later. */
   Curl_uint_bset_remove(&multi->dirty, data->mid);
 
+  /* Reset the retry count at the start of each request.
+   * If the retry count is not reset, when the connection drops,
+   * it will not enter the retry mechanism on CONN_MAX_RETRIES + 1 attempts
+   * and will immediately throw
+   * "Connection died, tried CONN_MAX_RETRIES times before giving up".
+   * By resetting it here, we ensure each new request starts fresh. */
+  data->state.retrycount = 0;
+
```

Comment:
This doesn't seem like the correct place to reset the counter. The counter is meant to be zero for a transfer and then get bumped once for every time it tries a new connection so that it doesn't get stuck in trying new connections that fail at once for some reason.

I believe assigning it to zero here effectively takes away that ability.

I think putting it in `Curl_pretransfer` might be fine.